desc/integrals/bounce_integral.py

```diff
@@ -1,26 +1,725 @@
+    return FourierChebyshevSeries(
+        f=interp_rfft2(
+            # Interpolate to nodes in Clebsch space,
+            # which is not a tensor product node set in DESC space.
+            xq=nodes.reshape(*nodes.shape[:-3], M * N, 2),
+            f=f[..., jnp.newaxis, :, :],
+            axes=(-1, -2),
+        ).reshape(*nodes.shape[:-3], M, N),
+        domain=(0, 2 * jnp.pi),
+    )
+
+
+def _transform_to_clebsch_1d(grid, alpha, theta, B, N_B, is_reshaped=False):
+    """Transform to single variable Clebsch spectral domain.
+
+    Parameters
+    ----------
+    grid : Grid
+        Tensor-product grid in (θ, ζ) with uniformly spaced nodes in
+        (2π × 2π) poloidal and toroidal coordinates.
+        Preferably power of 2 for ``grid.num_theta`` and ``grid.num_zeta``.
+    alpha : jnp.ndarray
+        Shape (L, num_transit) or (num_transit, ).
+        Sequence of poloidal coordinates A = (α₀, α₁, …, αₘ₋₁) that specify field line.
+    theta : jnp.ndarray
+        Shape (L, M, N) or (M, N).
+        DESC coordinates θ sourced from the Clebsch coordinates
+        ``FourierChebyshevBasis.nodes(M,N,domain=(0,2*jnp.pi))``.
+    B : jnp.ndarray
+        |B| evaluated on ``grid``.
+    N_B : int
+        Desired Chebyshev spectral resolution for |B|. Preferably power of 2.
+
+    Returns
+    -------
+    T, B : (PiecewiseChebyshevSeries, PiecewiseChebyshevSeries)
+        Set of 1D Chebyshev spectral coefficients of |B| along field line.
+        {|B|_α : ζ ↦ |B|(α, ζ) | α ∈ A } where A = (α₀, α₁, …, αₘ₋₁) is the
+        sequence of poloidal coordinates that specify the field line.
+        Likewise with θ.
+
+    """
+    if not is_reshaped:
+        B = grid.meshgrid_reshape(B, "rtz")
+
+    # Evaluating set of single variable maps is more efficient than evaluating
+    # multivariable map, so we project θ to a set of Chebyshev series.
+    T = FourierChebyshevSeries(f=theta, domain=(0, 2 * jnp.pi)).compute_cheb(alpha)
+    theta = T.evaluate(N_B)
+    xq = jnp.stack(
+        [theta, jnp.broadcast_to(cheb_pts(N_B, domain=T.domain), theta.shape)], axis=-1
+    ).reshape(*alpha.shape[:-1], alpha.shape[-1] * N_B, 2)
+    B = interp_rfft2(
+        xq=xq,
+        f=B[..., jnp.newaxis, :, :],
+        axes=(-1, -2),
+    ).reshape(*alpha.shape, N_B)
+    # Need |B| parameterized by single variable to compute roots.
+    B = PiecewiseChebyshevSeries(cheb_from_dct(dct(B, type=2, axis=-1)) / N_B, T.domain)
+    return T, B
+
+
+def _swap_pl(f):
+    # Given shape (L, num_pitch, -1) or (num_pitch, L, -1) or (num_pitch, -1)
+    # swap L and num_pitch axes.
+    assert f.ndim <= 3
+    return jnp.swapaxes(f, 0, -2)
+
+
+# TODO: After GitHub issue #1034 is resolved, we should pass in the previous
+#  θ(α, ζ) coordinates as an initial guess for the next coordinate mapping.
+#  Perhaps tell the optimizer to perturb the coefficients of the
+#  θ(α, ζ) directly? think this is equivalent to perturbing lambda.
+
+
+class Bounce2D(IOAble):
```

Comment:
By the way, what Neo does for their function approximation is to use 200x200 2D cubic spline. In chose 2D pseudospectral function approximation in `Bounce2D` from the 1D cubic spline function approximation in `Bounce1D` because

1. These numbers are large enough that FFTs will have significant performance improvements
2. Spectral interpolation of smooth functions is more efficient, so we can reduce desired `knots_per_transit` from 200 to 64 or potentially 32. The degree of the polynomial is then reduced from 600 to 64 (or 32), so we reduce the number of bounce integrals to compute by a factor 10 to 20.
3. With cubic spline interpolation, even at large knot density, the local maxima of |B| are harder to approximate compared to pseudospectral. That may be important for things like Gamma_c where the result strongly depends on the how well the derivative of |B| is approximated near the local maxima.

A negative is that root finding is theoretically more expensive, but there are methods to resolve this if that becomes a performance limitation.